googleapiclient/http.py

```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
     resp, content - Response from the http request (may be HTTP 5xx).
   """
   resp = None
+  content = None
+  exception = None
   for retry_num in range(num_retries + 1):
     if retry_num > 0:
-      sleep(rand() * 2**retry_num)
-      LOGGER.warning(
-          'Retry #%d for %s: %s %s%s' % (retry_num, req_type, method, uri,
-          ', following status: %d' % resp.status if resp else ''))
+      # Sleep before retrying.
+      sleep_time = rand() * 2 ** retry_num
+      LOGGER.warning('Sleeping %.2f seconds before retry %d of %d for %s: '
+                      '%s %s, following %s',
+                      sleep_time, retry_num, num_retries, req_type, method,
+                      uri, resp.status if resp else exception)
+      sleep(sleep_time)
 
     try:
+      exception = None
       resp, content = http.request(uri, method, *args, **kwargs)
-    except ssl.SSLError:
+    # Retry on SSL errors and socket timeout errors.
+    except ssl.SSLError as ex:
+      exception = ex
+    except socket.error as ex:
+      try:
+        # errno's contents differ by platform, so we have to match by name.
+        if (socket.errno.errorcode[ex.errno] not in
+            ('WSAETIMEDOUT', 'ETIMEDOUT')):
+          raise
+      # If we couldn't match it - retry because we can't know for sure.
+      except KeyError:
+        pass
+      exception = ex
+
+    if exception:
       if retry_num == num_retries:
-        raise
+        raise exception
       else:
         continue
-    if resp.status < 500:
+
+    if not _should_retry_response(resp.status, content):
       break
 
   return resp, content
 
 
+def _should_retry_response(resp_status, content):
+  """Determines whether a response justifies a retry.
+
+  Args:
+    resp_status: The response status received.
+    content: The response content body. 
+
+  Returns:
+    True if the response needs to be retried, otherwise False.
```

Comment:
In the space of ten lines you equate between "should", "justifies", and "needs" - pick one? Probably "should".